test/data/schema.js

```diff
@@ -0,0 +1,317 @@
+  },
+  outputFields: {
+    viewer: {
+      type: GraphQLUser,
+      resolve: getViewer,
+    },
+    changedTodos: {
+      type: new GraphQLList(GraphQLTodo),
+      resolve: ({ changedTodoIds }) => changedTodoIds.map(getTodo),
+    },
+  },
+  mutateAndGetPayload: ({ complete }) => {
+    const changedTodoIds = markAllTodos(complete);
+    return { changedTodoIds };
+  },
+});
+
+const GraphQLRemoveCompletedTodosMutation = mutationWithClientMutationId({
+  name: 'RemoveCompletedTodos',
+  outputFields: {
+    viewer: {
+      type: GraphQLUser,
+      resolve: getViewer,
+    },
+    deletedIds: {
+      type: new GraphQLList(GraphQLString),
+      resolve: ({ deletedIds }) => deletedIds,
+    },
+  },
+  mutateAndGetPayload: () => {
+    const deletedTodoIds = removeCompletedTodos();
+    const deletedIds = deletedTodoIds.map(toGlobalId.bind(null, 'Todo'));
+    return { deletedIds };
+  },
+});
+
+const GraphQLRemoveTodoMutation = mutationWithClientMutationId({
+  name: 'RemoveTodo',
+  inputFields: {
+    id: { type: new GraphQLNonNull(GraphQLID) },
+  },
+  outputFields: {
+    viewer: {
+      type: GraphQLUser,
+      resolve: getViewer,
+    },
+    deletedId: {
+      type: GraphQLID,
+      resolve: ({ id }) => id,
+    },
+  },
+  mutateAndGetPayload: ({ id }) => {
+    const { id: todoId } = fromGlobalId(id);
+    removeTodo(todoId);
+    return { id };
+  },
+});
+
+const GraphQLRenameTodoMutation = mutationWithClientMutationId({
+  name: 'RenameTodo',
+  inputFields: {
+    id: { type: new GraphQLNonNull(GraphQLID) },
+    text: { type: new GraphQLNonNull(GraphQLString) },
+  },
+  outputFields: {
+    todo: {
+      type: GraphQLTodo,
+      resolve: ({ todoId }) => getTodo(todoId),
+    },
+  },
+  mutateAndGetPayload: ({ id, text }) => {
+    const { id: todoId } = fromGlobalId(id);
+    renameTodo(todoId, text);
+    return { todoId };
+  },
+});
+
+const GraphQLMutation = new GraphQLObjectType({
+  name: 'Mutation',
+  fields: {
+    addTodo: GraphQLAddTodoMutation,
+    changeTodoStatus: GraphQLChangeTodoStatusMutation,
+    markAllTodos: GraphQLMarkAllTodosMutation,
+    removeCompletedTodos: GraphQLRemoveCompletedTodosMutation,
+    removeTodo: GraphQLRemoveTodoMutation,
+    renameTodo: GraphQLRenameTodoMutation,
+  },
+});
+
+// const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
+
+const GraphQLTodoUpdatedSubscription = subscriptionWithClientId({
+  name: 'TodoUpdatedSubscription',
+
+  inputFields: {
+    id: { type: new GraphQLNonNull(GraphQLID) },
+  },
+
+  outputFields: {
+    todo: {
+      type: GraphQLTodo,
+      resolve: ({ id }) => {
+        return getTodo(id);
+      },
+    },
+  },
+
+  // subscribe: ({ id }, { subscribe }) => {
+  //   return subscribe(`todo:${id}:updated`);
+  // },
+
+  async *subscribe({ id }, { subscribe }) {
+    console.log('subscribe', id);
+    const stream = await subscribe(`todo:${id}:updated`);
+
+    // await delay(50);
+    yield* stream;
+  },
```

Comment:
It might appear as if this could was functionally identical to the commented code above, however, this seems to be important for reproducing. I cannot reproduce this even with 1000 iterations with the commented code above. Perhaps Graphql handles the promise return differently? or more lazily? fully unsure at this point. This feels like we're dealing with races on the single event tick scale

Reply:
OK WE FIGURED THIS OUT. 

The issue isn't the `async` it's the `*`, the generator changes the behavior subtly. The code in the function body only runs when `.next()` is called, which happens on a different tick, causing it to be ordered after a unsubscribed on the same tick